### ISSUE TITLE
Analyzer: Warn when async assertions are not being awaited/stored

### ISSUE THREAD

**Issue by bradwilson** [xunit/xunit#2346]
### Discussed in https://github.com/xunit/xunit/discussions/2344

<sup>Originally posted by **leonardochaia** July  7, 2021</sup>

<blockquote>

Hey all!

We've been using `xUnit` on .NET 5 for a while now and we are loving it.

Some thing that I have noticed on junior devs that are just getting started, is that they forget to `await` the `Assert.ThrowsAsync` method.

For example:

```csharp
[Fact]
public void ThrowsAsync_ShouldWork()
{
    Task Callback() => throw new Exception();
    Assert.ThrowsAsync<InvalidOperationException>(Callback);
}
``` 

Is there a chance that the xUnit analyzer can warn about this issues? The example is giving a false positive.

If we await, the test fails:

```csharp
[Fact]
public async Task ThrowsAsync_ShouldWork()
{
    Task Callback() => throw new Exception();
    await Assert.ThrowsAsync<InvalidOperationException>(Callback);
}
```

So, I think it would be nice if the analyzer would warn about this cases, since the included .NET analyzers can't really know that you don't wanna do anything with the resulting Task.
</blockquote>

### Additional notes

I agree that we should have analyzers that inspect all async assertions (not just `ThrowsAsync`), and we should check for either `await` or assignment of the task object into a variable so that we know the user has explicitly chosen how to deal with the async assertion (if it's not too hard, throwaways are likely to be a bug, so maybe this should be an error if we don't have any `await` or assignment, and a warning if the assignment is to a throwaway variable).

**Comment by slaneyrw** [comment]
Is this just as "simple" as checking the Parent of the invocationOperation of of Kind OperationKind.Await.  If so, I can put in checks for 
ThrowsAsync, ThrowsAnyAsync, RaisesAsync, RaisesAnyAsync, and PropertyChangedAsync.

IMHO, this should be an error, not a warning though!

**Comment by bradwilson** [comment]
Definitely agree that it should be an error. If you don't await the `Task` (or store it away for later awaiting) you are gonna have a bad time.

**Comment by bradwilson** [comment]
@bjornhellander This is a "pit of success" thing: our analyzers are on by default for most users of xUnit.net, whereas the analyzers you suggest are not on by default for anybody. This is very specifically about the assertions that are async, not generalized "failure to await", because it can cause false positives when testing.

**Comment by bjornhellander** [comment]
Ok. I just thought it would be a waste of time to re-implement something that already exists, but I see your point.

**Comment by bradwilson** [comment]
Available in `1.5.0-pre.15`.
